File: packages/node/src/Seguranca/Ataques/AtaqueIdaEVoltaBinario.js
'use strict';

const ResultadoAtaque = require('../ResultadoAtaque.js');
const { randomBytes } = require('../Util.js');

/**
 * O AtaqueIdaEVolta usa só o caractere 'Q'. Esse usa dados binários de
 * verdade: todos os 256 valores de byte, NUL, sequências aleatórias de
 * vários tamanhos. Erros de manipulação de string (trim, encoding, NUL
 * truncation) só aparecem com bytes arbitrários.
 *
 * Adaptação Node: os dados são Buffers; como decrypt() devolve string
 * (UTF-8), comparamos com a representação UTF-8 dos mesmos bytes.
 */
class AtaqueIdaEVoltaBinario {
  constructor({ tamanhoMaximo = 256 } = {}) {
    this.tamanhoMaximo = tamanhoMaximo;
  }

  nome() {
    return 'Ida-e-volta com dados binários (inclui NUL)';
  }

  executar(alvo) {
    const falhas = [];

    for (let len = 0; len <= this.tamanhoMaximo; len++) {
      const texto = len === 0 ? Buffer.alloc(0) : randomBytes(len);
      try {
        if (alvo.decrypt(alvo.encrypt(texto)) !== texto.toString('utf8')) {
          falhas.push(len);
        }
      } catch (e) {
        falhas.push(`${len} (erro: ${e.message})`);
      }
    }

    const todos = Buffer.from(Array.from({ length: 256 }, (_, i) => i));
    if (alvo.decrypt(alvo.encrypt(todos)) !== todos.toString('utf8')) {
      falhas.push('todos os 256 valores de byte');
    }

    if (falhas.length > 0) {
      return new ResultadoAtaque(
        this.nome(),
        true,
        'critica',
        'Falhou em ' + falhas.length + ' caso(s): ' + falhas.slice(0, 10).join(', '),
        { falhas },
      );
    }

    return new ResultadoAtaque(
      this.nome(),
      false,
      'info',
      `Todos os tamanhos 0..${this.tamanhoMaximo} e os 256 valores de byte preservados`,
    );
  }
}

module.exports = AtaqueIdaEVoltaBinario;

File: packages/dart/lib/Seguranca/Ataques/AtaqueIdaEVoltaBinario.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:criptografia/Seguranca/AlvoCriptografico.dart';
import 'package:criptografia/Seguranca/Ataque.dart';
import 'package:criptografia/Seguranca/ResultadoAtaque.dart';
import 'package:criptografia/Seguranca/Util.dart';

/// O AtaqueIdaEVolta usa só o caractere 'Q'. Esse usa dados binários de
/// verdade: todos os 256 valores de byte, NUL, sequências aleatórias de
/// vários tamanhos. Erros de manipulação de string (trim, encoding, NUL
/// truncation) só aparecem com bytes arbitrários.
///
/// Adaptação Dart: os dados são bytes; como decrypt() devolve string
/// (UTF-8, com bytes inválidos virando U+FFFD), comparamos com a representação
/// UTF-8 leniente dos mesmos bytes.
class AtaqueIdaEVoltaBinario extends Ataque {
  final int _tamanhoMaximo;

  AtaqueIdaEVoltaBinario({this._tamanhoMaximo = 256});

  @override
  String nome() => 'Ida-e-volta com dados binários (inclui NUL)';

  @override
  ResultadoAtaque executar(AlvoCriptografico alvo) {
    final falhas = <String>[];

    for (var len = 0; len <= _tamanhoMaximo; len++) {
      final texto = len == 0 ? Uint8List(0) : randomBytes(len);
      try {
        if (alvo.decrypt(alvo.encrypt(texto)) !=
            utf8.decode(texto, allowMalformed: true)) {
          falhas.add('$len');
        }
      } catch (e) {
        falhas.add('$len (erro: $e)');
      }
    }

    final todos = Uint8List.fromList(List<int>.generate(256, (i) => i));
    if (alvo.decrypt(alvo.encrypt(todos)) !=
        utf8.decode(todos, allowMalformed: true)) {
      falhas.add('todos os 256 valores de byte');
    }

    if (falhas.isNotEmpty) {
      return ResultadoAtaque(
        nome(),
        true,
        Severidade.critica,
        'Falhou em ${falhas.length} caso(s): ${falhas.take(10).join(', ')}',
        {'falhas': falhas},
      );
    }

    return ResultadoAtaque(
      nome(),
      false,
      Severidade.info,
      'Todos os tamanhos 0..$_tamanhoMaximo e os 256 valores de byte '
          'preservados',
    );
  }
}
